### ISSUE TITLE
Walgreens block

### ISSUE THREAD

**Issue by paf0186** [GUI/covid-vaccine-spotter#74]
There's a commit referring to a Walgreens block (changing the user agent to Firefox, rather than a special one), but it seems it's blocked again.

Is there anything a person can do to help?  Details of the issue or, etc?  (I'm guessing not, but thought I'd offer.)

**Comment by GUI** [comment]
Thanks! Yeah, unfortunately it seems like Walgreens has blocked things. Last night they started blocking things, but this was resolved by changing the user agent. But today's block seems slightly different. I unfortunately haven't had a chance to really dig into this yet, but it seems like it's not just a matter of user agent, HTTP headers, or IP addresses. I can make a few requests to the timeslots endpoint but then I start getting 403 Forbidden responses across the board. I'm unfortunately going to have to punt this to tomorrow, but I'll definitely try to dig into this and get it fixed. But if anyone else has done any further debugging or has fixes, I'm definitely open to ideas or contributions. In the meantime, sorry!

**Comment by ontanedf** [comment]
I just wanted to say I don't have enough know-how to be able to help with this issue from a coding side but I will gladly be an agent of trying to get Walgreens to unblock by contacting them/posting on social media and trying to get them to unblock!

**Comment by brianlong** [comment]
@GUI You are welcome to use my proxy network, at no charge, for this project. Please look for emails from me or DM me (I am also @brianlong on Twitter).

**Comment by paf0186** [comment]
I have to say, it sounds like a good chance to be rate limiting, so Brian's offer is *very* welcome.

**Comment by jameskip** [comment]
Hey, james with texasupdates.com here

Hopefully, Walgreens will allow access again soon. Has there been any communication with Walgreens?

**Comment by james-horton** [comment]
I believe the block is cookie related because I bookmarked the search page and started to get an access denied recently. Only after going to the landing page first do you get the necessary cookies to view the search page. It would be sweet if Walgreens put out an API and fully opened it up.

**Comment by paf0186** [comment]
James,

Could you give the specific URLs of the search and landing pages?  Thanks!

**Comment by james-horton** [comment]
**Landing**
https://www.walgreens.com/findcare/vaccination/covid-19
**Search**
https://www.walgreens.com/findcare/vaccination/covid-19/location-screening

**Comment by jeffchenoweth** [comment]
I ran into this with my scraper. Got banned and couldn’t get to the landing and search pages from chrome. Not sure what dynatrace is but the keep track of how many requests you send to the appointments endpoint in a cookie.

**Comment by GUI** [comment]
I've very tentatively got Walgreens working again, but this is of course subject to change.

@brianlong: I just followed up via e-mail, but thank you so much for the offer! Right now, I think I'm actually set for proxies, but I'll be sure to reach out if anything changes. Thank you!!

In terms of details, here was the potential/hopeful fix: https://github.com/GUI/covid-vaccine-spotter/commit/ced1d601a23121c39836ca4dd448ed8e4a4c9276 One part was specific to my implementation (and appears to maybe be some super weird issue leading Akamai to blocking Node.js based HTTP clients, or I'm missing something obvious), the other issue was more related to rate limits as some of you all have seen. The longer explanation:

> 1. It seems like we're getting somehow blocked when making requests via Node.js HTTP clients, like "got". I saw this earlier with some random Southern Grocers requests, but it's all very strange. I'm either missing something obvious, or this is some more mysterious Node.js bug. Identical HTTP headers fail when using a nodejs native client vs curl. But if we use a curl client, then things seem to work. Even stranger, the Node.js clients seem to start working if I route them through a proxy to observe (like mitmproxy), so again, it's all very strange (or I'm missing something).
> 
>    I still want to investigate this further, but I believe the blocked requests on Walgreens and Southern Grocers are maybe both coming from an Akamai layer, so I don't know if somehow Akamai is picking up on some odd Node.js-specific behavior and rejecting. Or I'm missing something obvious.
> 2. In any case, with that fix in place, Walgreens now seems to be imposing more strict IP rate limits for requests. But I can't exactly figure out the rhyme or reason to the limits, since they sometimes seem to still crop up randomly even if I slow things way down.
> 
>    So I'm not wild about this workaround, but I'm using a proxy service to route requests across a variety of random IPs. I really don't want to evade any purposeful rate limits Walgreens might be adding to deal with load issues, but I'm also not sure if these are intentional or just part of some other Akamai-level thing. I originally used a User-Agent that would clearly identify my bot (and point to my github repo), so I would definitely welcome any feedback from Walgreens on acceptable rates. But in light of Walgreens saying they support these appointment finding initiatives (https://www.bloomberg.com/news/articles/2021-03-16/covid-vaccine-signup-woes-prompting-a-wave-of-web-workarounds), I'm going to go forward with this workaround unless I hear that this is not acceptable.

**Comment by MoralCode** [comment]
This seems like it might be resolved?

Would it be helpful to compile all the notes on various combinations of scraping techniques that people have tried into a wiki on this repo so people can find that info in one place?

**Comment by carefulcomputer** [comment]
i observed that script is not fetching location from SF Bay area. Any possibility that it is not checking zip codes from bay area (e.g. San Francisco, Redwood City, Fremont - 94541, 94121 ). Zipcodes outside bay area work fine (e.g. Modesto, Stockton - 95351  etc)

**Comment by GUI** [comment]
@carefulcomputer: I'm not entirely sure it was affecting that location, but I did discover some edge-cases where Walgreens data might not be getting updated depending on how many locations were in an area. This was largely addressed in https://github.com/GUI/covid-vaccine-spotter/commit/0a528cc6a70a5622f53e2a7910cb6ab5be2c6a6d

I've also tried to further optimize how Walgreens searches are performed, which should cut down on the number of requests needed, which should speed things up a bit and hopefully prevent further blocks. https://github.com/GUI/covid-vaccine-spotter/commit/ca8da7fafaa6e1b28c373639d5e3bfc1d195b4e0

So I'm going to close this as resolved now that I think the Walgreens implementation is working a bit better now. Hopefully that won't jinx things and we won't get blocked again. :)
